test/simulator/main/org/apache/cassandra/simulator/paxos/StrictSerializabilityValidator.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.simulator.paxos;
+
+import javax.annotation.Nullable;
+
+import accord.verify.StrictSerializabilityVerifier;
+import com.carrotsearch.hppc.IntIntHashMap;
+import com.carrotsearch.hppc.IntIntMap;
+
+public class StrictSerializabilityValidator implements HistoryValidator, HistoryValidator.Checker
+{
+    private final StrictSerializabilityVerifier verifier;
+    private final IntIntMap index;
+    private Observation observation;
+
+    public StrictSerializabilityValidator(int[] primaryKeys)
+    {
+        this.verifier = new StrictSerializabilityVerifier(primaryKeys.length);
+        index = new IntIntHashMap(primaryKeys.length);
+        for (int i = 0; i < primaryKeys.length; i++)
+            index.put(primaryKeys[i], i);
+    }
+
+    @Override
+    public Checker start(Observation observation)
+    {
+        this.observation = observation;
+        verifier.begin();
+        return this;
+    }
+
+    @Override
+    public void log(@Nullable Integer pk)
+    {
+        if (pk == null) verifier.print();
+        else verifier.print(get(pk));
+    }
+
+    private int get(int pk)
+    {
+        if (index.containsKey(pk))
+            return index.get(pk);
+        throw new IllegalArgumentException("Unknown pk=" + pk);
+    }
+
+    @Override
+    public void witnessRead(int pk, int count, int[] seq)
+    {
+        convertHistoryViolation(() -> verifier.witnessRead(get(pk), seq));
```

Comment:
I don't believe this call can throw a violation? Nor the write one.

Reply:
it doesn't now, but figured it "may" in the future, so safer to guard against

Reply:
I think that is actively confusing. It makes the reader think the method can throw, which is surprising if you understand the API involves building a collection of changes to apply together. Defensive programming is good, but only if it communicates the right thing to the reader.

Reply:
removed